plugins/org.eclipse.fordiac.ide.application/src/org/eclipse/fordiac/ide/application/marker/resolution/FordiacMarkerResolutionGenerator.java

```diff
@@ -0,0 +1,66 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Primetals Technologies Austria GmbH
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Sebastian Hollersbacher
+ *    - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+package org.eclipse.fordiac.ide.application.marker.resolution;
+
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.fordiac.ide.model.errormarker.FordiacErrorMarker;
+import org.eclipse.fordiac.ide.model.libraryElement.FB;
+import org.eclipse.fordiac.ide.model.libraryElement.LibraryElementPackage;
+import org.eclipse.fordiac.ide.model.typelibrary.impl.ErrorFBTypeEntryImpl;
+import org.eclipse.ui.IMarkerResolution;
+import org.eclipse.ui.IMarkerResolutionGenerator2;
+
+public class FordiacMarkerResolutionGenerator implements IMarkerResolutionGenerator2 {
+
+	@Override
+	public IMarkerResolution[] getResolutions(final IMarker marker) {
+		try {
+			final EClass targetType = FordiacErrorMarker.getTargetType(marker);
+			if (LibraryElementPackage.Literals.ERROR_MARKER_FBN_ELEMENT.equals(targetType)
+					|| LibraryElementPackage.Literals.FB.equals(targetType)
+							&& FordiacErrorMarker.getTarget(marker) instanceof final FB fb
+							&& fb.getTypeEntry() instanceof ErrorFBTypeEntryImpl) {
+				return new IMarkerResolution[] { new CreateMissingFBMarkerResolution(marker),
+						new ChangeFBMarkerResolution(marker) };
+			}
+			if (LibraryElementPackage.Literals.VAR_DECLARATION.equals(targetType)) {
+				return new IMarkerResolution[] { new ChangeDataTypeMarkerResolution(marker),
+						new CreateDataTypeMarkerResolution(marker) };
+			}
+		} catch (final IllegalArgumentException | CoreException e) {
+			e.printStackTrace();
+		}
+
+		return new IMarkerResolution[0];
+	}
+
+	@Override
+	public boolean hasResolutions(final IMarker marker) {
+		try {
+			final EClass targetType = FordiacErrorMarker.getTargetType(marker);
+			if (LibraryElementPackage.Literals.FB.equals(targetType)
+					&& FordiacErrorMarker.getTarget(marker) instanceof final FB fb
+					&& fb.getTypeEntry() instanceof ErrorFBTypeEntryImpl) {
+				return true;
+			}
+
+			return LibraryElementPackage.Literals.ERROR_MARKER_FBN_ELEMENT.equals(targetType)
+					|| LibraryElementPackage.Literals.VAR_DECLARATION.equals(targetType);
+		} catch (final IllegalArgumentException | CoreException e) {
+			return false;
+		}
```

Comment:
I would suggest to filter using the new `FordiacErrorMarker.getSource/Code(IMarker)` from PR https://github.com/eclipse-4diac/4diac-ide/pull/421 and avoid `getTarget(IMarker)` altogether.